### DISCUSSION TITLE
How can I simulate the operation of this library?

### DISCUSSION THREAD

**Discussion by viniciusbaltoe** [Open-Agriculture/AgIsoStack-plus-plus#271] [Q&A] [answered]
Hi, I'm an electrical engineering student from Brazil and I'm interested in learning more about ISOBUS and its surroundings. Unfortunately, documentation of ISOBUS comes at a pretty high price for me, so I'm looking for ways to learn about it from information I find on the internet. Observing AgIsoStack++, I was able to understand some things, but I don't have, at this point, equipment to prepare tests and carry out the tutorial. So, do you have any suggestions for some form of simulation that I can run?

**Comment by ad3154** [comment, marked as answer]
Hello! It depends on what exactly you want to simulate.

If you're looking to essentially make two applications that talk to each other via ISOBUS without any hardware, that's easy enough if you have access to a Linux PC of some kind, like a raspberry pi, or a desktop PC with Linux installed. Basically, you could set up two programs to communicate over Linux virtual CAN interfaces.

```
sudo modprobe vcan
sudo ip link add dev vcan0 type vcan
sudo ip link set up vcan0
```

Then when running our tutorials/examples, just replace "can0" with "vcan0" when using the SocketCAN driver.

`canDriver = std::make_shared<isobus::SocketCANInterface>("vcan0");`

If you're wanting to do something a bit more complicated, like perform virtual terminal or task controller communication, that might be pretty difficult, as we do not have software to fully emulate a VT or TC (yet) unfortunately. Those will probably require some hardware to fully experiment with.
